### PR TITLE
Fix Streaming#memoization.

### PR DESCRIPTION
**Pull request by non**
This commit fixes a bug noticed by @ceedubs. The basic issue
is that we have to be sure to memoize the Eval instance holding
the tail while *also* memoizing the tail itself when it is
calculated. Previously we were only memoizing the head node.

### CODE REVIEW

File: core/src/main/scala/cats/data/Streaming.scala [hunk 1]
```diff
@@ -642,14 +642,21 @@ sealed abstract class Streaming[A] extends Product with Serializable { lhs =>
...
-   * head of the stream is retained.
+   * By default this structure does not memoize to avoid memory leaks
+   * when the head of the stream is retained. However, the user
+   * ultimately has control of the memoization approach based on what
+   * kinds of Eval instances they use.
+   *
+   * There are two calls to memoized here -- one is a recursive call
+   * to this method (on the tail) and the other is a call to memoize
+   * the Eval instance holding the tail. For more information on how
+   * .memoize works see Eval#memoize.
```
Comment (ceedubs) [inline core/src/main/scala/cats/data/Streaming.scala]: Minor: could we turn this into a proper ScalaDoc link? `[[cats.Eval.memoize]]`

**Comment by ceedubs** [comment]
This fixes (and replaces) #662.

I left a minor request about the ScalaDoc, but :+1: from me.

File: core/src/main/scala/cats/data/Streaming.scala (hunk 1 above)
Comment (non) [inline core/src/main/scala/cats/data/Streaming.scala]: Oh, sure.

File: tests/src/test/scala/cats/tests/StreamingTests.scala [hunk 2]
```diff
@@ -37,6 +37,28 @@ class StreamingTests extends CatsSuite {
...
+      xs should === (ns)
+
+      val ys = memoized.toList
+      i should === (size)
+      ys should === (ns)
+    }
+  }
+
+  // convert List[A] to Streaming[A]
+  def convert[A](as: List[A]): Streaming[A] =
```
Comment (adelbertc) [inline tests/src/test/scala/cats/tests/StreamingTests.scala]: Is this used anywhere, or just added for possible future use?

**Comment by adelbertc** [comment]
:+1: once question about function is answered :-)

File: tests/src/test/scala/cats/tests/StreamingTests.scala (hunk 2 above)
Comment (non) [inline tests/src/test/scala/cats/tests/StreamingTests.scala]: Oh, looks like I left that in. I'll remove it.

**Comment by adelbertc** [comment]
Strange error, rebuilding, once it's green :+1:

**Pull request merged**